core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala

```diff
@@ -274,40 +260,43 @@ trait SubmissionMonitor extends FutureSupport with LazyLogging with RawlsInstrum
       logger.error(s"Failure monitoring workflow in submission $submissionId", t)
     }
 
-    //Update the workflow statuses in their own transaction
-    val workflowsWithOutputsF = datasource.inTransaction { dataAccess =>
-      val updatedRecs = response.statusResponse.collect {
-        case Success(Some((updatedRec, _))) => updatedRec
+    val workflowsWithOutputs = response.statusResponse.collect {
+      case Success(Some((workflowRec, Some(outputs)))) => (workflowRec, outputs)
+    }
+
+    // Attach the outputs in a txn of their own. If attaching outputs fails, it will mark the workflow as failed. This is fine.
+    val attachOutputsTxn = datasource.inTransaction { dataAccess =>
+      handleOutputs(workflowsWithOutputs, dataAccess)
+    }
+
+    // Update statuses for workflows and submission in a separate txn.
+    val statusUpdateTxn = datasource.inTransaction { dataAccess =>
+      // Refetch workflows as some may have been marked as Failed by handleOutputs.
+      val workflowUpdates = dataAccess.workflowQuery.findWorkflowByIds( workflowsWithOutputs.map( _._1.id ) ).result map { updatedRecs =>
+        //New statuses according to the execution service.
+        val workflowIdToNewStatus = workflowsWithOutputs.map({ case (workflowRec, outputs) => workflowRec.id -> workflowRec.status }).toMap
```

Comment:
outputs are not used here, so you can do `case (workflowRec, _)` or even just get the `keys()` first